sdw_updater/Updater.py

```diff
@@ -104,40 +106,124 @@ def migration_is_required():
-            else:
-                upgrade_results = UpdateStatus.UPDATES_OK
+
+def apply_updates_templates(templates=current_templates, progress_callback=None):
+    """
+    Apply updates to all TemplateVMs.
+    """
+    sdlog.info("Applying all updates to VMs: {}".format(", ".join(templates)))
+    try:
+        proc = _start_qubes_updater_proc(templates)
+        result_update_status = {}
+        stderr_thread = threading.Thread(
+            target=_qubes_updater_parse_progress,
+            args=(proc.stderr, result_update_status, templates, progress_callback),
+        )
+        stdout_thread = threading.Thread(target=_qubes_updater_parse_stdout, args=(proc.stdout,))
+        stderr_thread.start()
+        stdout_thread.start()
+
+        while proc.poll() is None or stderr_thread.is_alive() or stdout_thread.is_alive():
+            time.sleep(1)
+
+        stderr_thread.join()
+        stdout_thread.join()
+        proc.stderr.close()
+        proc.stdout.close()
+        update_status = overall_update_status(result_update_status)
+        if update_status == UpdateStatus.UPDATES_OK:
+            sdlog.info("Template updates successful")
         else:
-            upgrade_results = _apply_updates_vm(vm)
+            sdlog.info("Template updates failed")
+        return update_status
+
+    except subprocess.CalledProcessError as e:
+        sdlog.error(
+            "An error has occurred updating templates. Please contact your administrator."
+            " See {} for details.".format(DETAIL_LOG_FILE)
+        )
+        sdlog.error(str(e))
+        return UpdateStatus.UPDATES_FAILED
+
 
-        progress_current += progress_step
+def _start_qubes_updater_proc(templates):
+    update_cmd = [
+        "qubes-vm-update",
+        "--restart",  # Enforce app qube restarts
+        "--show-output",  # Update transaction details (goes to stdout)
+        "--just-print-progress",  # Progress reporting (goes to stderr)
+        "--targets",
+        ",".join(templates),
+    ]
+    detail_log.info("Starting Qubes Updater with command: {}".format(" ".join(update_cmd)))
+    return subprocess.Popen(
+        update_cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
 
-        # constrain progress_current to given progress range
-        def clamp(val, min_val, max_val):
-            return max(min(max_val, val), min_val)
 
-        progress_current = clamp(progress_current, progress_start, progress_end)
+def _qubes_updater_parse_stdout(stream):
+    while True:
+        untrusted_line = stream.readline()
+        if len(untrusted_line) == 0:
+            break
 
-        yield vm, progress_current, upgrade_results
+        line = Util.strip_ansi_colors(untrusted_line.decode("utf-8"))
+        line = line.rstrip()
+        detail_log.info("[Qubes updater] {}".format(line))
+
+
+def _qubes_updater_parse_progress(stream, result, templates, progress_callback=None):
```

Comment:
It's so much more user-friendly now with the detailed VM progress reporting- I know it was a bit of a pain but thanks for making these changes. :)